services/stats/stats.go

```diff
@@ -93,29 +97,66 @@ type RudderStatsT struct {
 	dontProcess bool
 }
 
+func getNewStatsdClientWithExpoBackoff(opts ...statsd.Option) {
+	maxWait := time.Minute * 10
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxInterval = time.Minute
+	bo.MaxElapsedTime = maxWait
+	var err error
+
+	getNewClient := func() error {
+		//statsd.New returns client even when err!=nil, although that client won't work.
+		//But, since we don't want getting new client to be a blocking call. So, setting it even with the non-working client.
+		//And, will be updated once we have `err==nil`, until then stats sent to telegraf will be dropped.
+		client, err = statsd.New(opts...)
+		if err != nil {
+			isMockClient = true
+			pkgLogger.Errorf("error while setting statsd client: %v", err)
+		}
+		return err
+	}
+
+	if err = backoff.Retry(getNewClient, bo); err != nil {
+		if bo.NextBackOff() == backoff.Stop {
+			//if setup didn't succeed in bo.MaxElapsedTime, then we panic. Since, if `RSERVER_ENABLE_STATS` is `true`
+			// & we failed to complete statsd setup in maxElapsedTime then something is wrong. And, we don't want to loose metrics by ignoring the error.
+			panic(err)
+		}
+	}
+}
+
 //Setup creates a new statsd client
 func Setup() {
 	DefaultStats = &HandleT{}
 
 	if !statsEnabled {
 		return
 	}
-
-	var err error
 	conn = statsd.Address(statsdServerURL)
-	//TODO: Add tags by calling a function...
+	// since, we don't want setup to be a blocking call, creating a separate `go routine`` for retry to get statsd client.
+	var err error
+	//NOTE: this is to get atleast a dummy client, even in case of failure. So, that nil pointer error is not received when client is called.
 	client, err = statsd.New(conn, statsd.TagsFormat(getTagsFormat()), defaultTags())
-	if err != nil {
-		// If nothing is listening on the target port, an error is returned and
-		// the returned client does nothing but is still usable. So we can
-		// just log the error and go on.
-		pkgLogger.Error(err)
-	}
-	if client != nil {
-		rruntime.Go(func() {
-			collectRuntimeStats(client)
-		})
-	}
+	rruntime.Go(func() {
+		if err != nil {
+
+			getNewStatsdClientWithExpoBackoff(conn, statsd.TagsFormat(getTagsFormat()), defaultTags())
```

Comment:
Let's rename this to avoid the deepsource error:
```
function 'getNewStatsdClientWithExpoBackoff' seems to be a getter but it does not return any result
```

Reply:
since we have global variable `client` that is being set by calling this function, converting `getNewStatsdClientWithExpoBackoff` to `setNewStatsdClientWithExpoBackoff`